### DISCUSSION TITLE
Is There a Way to Create a Thread Local Exclusive System?

### DISCUSSION THREAD

**Discussion by zicklag** [bevyengine/bevy#1636] [Q&A]
I need to create an exclusive system that is local to the main thread. The issue is that the system I have is `!Send + !Sync`. At the same time, I only need it to be thread local on WASM, but there aren't threads on WASM anyway right? Maybe I can just `unsafe impl Send {}` for the `!Send` because WASM doesn't have threads so it won't be a problem?

**Comment by Ratysz** [comment]
Exclusive systems always run on the main thread. Or are there API discrepancies that don't let it work as expected? Can you provide an example?

**Reply by zicklag** [reply]
Oh, OK. Here's an example that doesn't work:

```rust
use bevy::prelude::*;

struct NotSend {
    not_send: *const u8,
}

fn main() {
    let system = make_system();
    App::build().add_system(system.exclusive_system()).run();
}

fn make_system() -> impl Fn(&mut World) {
    let not_send = NotSend {
        not_send: std::ptr::null(),
    }

    move |_| {
        let _ = not_send;
    }
}
```

It complains that `exclusive_system` is not implemented for `system` in the current scope because the `IntoExclusiveSystem` trait requires that the `Fn` be `Sync + Send`:

https://github.com/bevyengine/bevy/blob/3a2a68852c0a1298c0678a47adc59adebe259a6f/crates/bevy_ecs/src/system/exclusive_system.rs#L43-L45

---

If exclusive systems only run on the current thread, then `Send + Sync` shouldn't be required for that impl should it?

**Reply by Ratysz** [reply]
Oh, I may have forgotten that closures actually work for exclusive systems...

Naively removing those bounds from exclusive systems and `Stage` seems to work, but I'm not clear on all the implications yet. Ideally, we'd like to allow things like `Local<(World, SystemStage)>` in parallel (and/or async) systems - it doesn't fully work or even make sense right now, but removing the bounds would be a step further away from it in either case.

Gonna need more thinking time to get a good solution.

**Reply by zicklag** [reply]
Yeah, for now, in my particular use-case, I was able to just `unsafe`-ly implement `Sync + Send` for struct which I only use in the WASM build anyway and I think that seems to work. So at least I'm not waiting on anything. :slightly_smiling_face:
